src/number_theory/modular_forms/basic.lean

```diff
@@ -0,0 +1,286 @@
+begin
+  simp only [weakly_modular_mem', pi.mul_apply, coe_coe] at *,
+  intros γ z,
+  rw [(hf γ z), (hg γ z)],
+  have pown := zpow_add₀ (upper_half_plane.denom_ne_zero (γ : GL(2, ℝ)⁺) z) k_1 k_2,
+  simp only [upper_half_plane.denom, coe_fn_coe_base, ne.def,
+    matrix.general_linear_group.coe_fn_eq_coe, coe_coe] at pown,
+  rw pown,
+  ring,
+end
+
+/-- A function `f : ℍ → ℂ` is a modular form weight `k ∈ ℤ` and of level `Γ` if it is holomorphic,
+ weakly modular and bounded at infinity. -/
+structure is_modular_form_of_weight_and_level (k : ℤ) (Γ : subgroup SL(2,ℤ)) (f : ℍ → ℂ) : Prop :=
+(hol : mdifferentiable 𝓘(ℂ) 𝓘(ℂ) (↑f : ℍ' → ℂ))
+(transf : f ∈ weakly_modular_submodule k Γ)
+(infinity : ∀ (A : SL(2,ℤ)), is_bounded_at_im_infty (f ∣[k, A]))
+
+/-- A function `f : ℍ → ℂ` is a cusp form of weight `k ∈ ℤ` and of level `Γ` if it is holomorphic,
+ weakly modular, and zero at infinity. -/
+structure is_cusp_form_of_weight_and_level (k : ℤ) (Γ : subgroup SL(2,ℤ)) (f : ℍ → ℂ) : Prop :=
+(hol : mdifferentiable 𝓘(ℂ) 𝓘(ℂ) (↑f : ℍ' → ℂ))
+(transf : f ∈ weakly_modular_submodule k Γ)
+(infinity : ∀ (A : SL(2,ℤ)), is_zero_at_im_infty (f ∣[k, A]))
+
+/-- The zero modular form is a cusp form-/
+lemma zero_cusp_form : is_cusp_form_of_weight_and_level k Γ 0 :=
+{ hol := by {apply mdifferentiable_zero,},
+  transf := (weakly_modular_submodule k Γ).zero_mem',
+  infinity := by { intro A,
+    rw slash_action.mul_zero,
+    apply (zero_at_im_infty_submodule ℂ).zero_mem}}
+
+lemma is_modular_form_of_weight_and_level_of_is_cusp_form_of_weight_and_level {f : ℍ → ℂ}
+  (h : is_cusp_form_of_weight_and_level k Γ f) : is_modular_form_of_weight_and_level k Γ f :=
+{ hol := h.1,
+  transf := h.2,
+  infinity := λ (A : SL(2,ℤ)), by {apply filter.zero_at_filter_is_bounded_at_filter _ _ (h.3 A)} }
+
+ /-- The zero modular form is a modular form-/
+lemma zero_mod_form : is_modular_form_of_weight_and_level k Γ 0 :=
+by apply_rules [is_modular_form_of_weight_and_level_of_is_cusp_form_of_weight_and_level,
+    zero_cusp_form]
+
+
+/-- This is the space of modular forms of weight `k` and level `Γ`-/
+def space_of_mod_forms_of_weight_and_level (k : ℤ) (Γ : subgroup SL(2,ℤ)) : submodule ℂ (ℍ → ℂ) :=
+{ carrier := { f : ℍ → ℂ | is_modular_form_of_weight_and_level k Γ f },
+  zero_mem':= by { simp only [set.mem_set_of_eq], apply zero_mod_form},
+  add_mem' := begin intros a b ha hb,
+    split,
+    { exact mdifferentiable_add ha.hol hb.hol },
+    { exact (weakly_modular_submodule k Γ).add_mem' ha.transf hb.transf} ,
+    { intro A,
+      rw slash_action.add_action,
+      exact (bounded_at_im_infty_subalgebra ℂ).add_mem' (ha.infinity A) (hb.infinity A)}
+    end,
+  smul_mem' := begin intros c f hf,
+    split,
+    { exact mdifferentiable_smul _ hf.hol },
+    { exact (weakly_modular_submodule k Γ).smul_mem' _ hf.transf },
+    { intro A,
+      rw slash_action.smul_action,
+      apply ((bounded_at_im_infty_subalgebra ℂ).smul_mem (hf.infinity A))}
+    end}
```

Comment:
```suggestion
  smul_mem' := λ c f hf,
  { hol := mdifferentiable_smul _ hf.hol,
    transf := (weakly_modular_submodule k Γ).smul_mem' _ hf.transf,
    infinity := λ A, begin
      rw slash_action.smul_action,
      apply (bounded_at_im_infty_subalgebra ℂ).smul_mem (hf.infinity A)
    end } }
```
Recommended, here and on the other fields of this one and the cusp forms -- it's more readable when the properties are proved by name rather than with a `split`.